plugins/linux-pipewire/camera-portal.c

```diff
@@ -0,0 +1,594 @@
+	uint32_t version;
+
+	ensure_camera_portal_proxy();
+
+	if (!camera_proxy)
+		return 0;
+
+	cached_version =
+		g_dbus_proxy_get_cached_property(camera_proxy, "version");
+	version = cached_version ? g_variant_get_uint32(cached_version) : 0;
+
+	return version;
+}
+
+/* ------------------------------------------------- */
+
+struct camera_device {
+	uint32_t id;
+	struct pw_properties *properties;
+	struct pw_proxy *proxy;
+	struct spa_hook proxy_listener;
+};
+
+static struct camera_device *
+camera_device_new(uint32_t id, const struct spa_dict *properties)
+{
+	struct camera_device *device = bzalloc(sizeof(struct camera_device));
+	device->id = id;
+	device->properties = properties ? pw_properties_new_dict(properties)
+					: NULL;
+	return device;
+}
+
+static void camera_device_free(struct camera_device *device)
+{
+	if (!device)
+		return;
+
+	g_clear_pointer(&device->proxy, pw_proxy_destroy);
+	g_clear_pointer(&device->properties, pw_properties_free);
+	bfree(device);
+}
+
+/* ------------------------------------------------- */
+
+static bool update_device_id(struct camera_portal_source *camera_source,
+			     const char *new_device_id)
+{
+	if (strcmp(camera_source->device_id, new_device_id) == 0)
+		return false;
+
+	g_clear_pointer(&camera_source->device_id, bfree);
+	camera_source->device_id = bstrdup(new_device_id);
+
+	return true;
+}
+
+static void stream_camera(struct camera_portal_source *camera_source)
+{
+	struct camera_device *device;
+
+	g_clear_pointer(&camera_source->obs_pw_stream,
+			obs_pipewire_stream_destroy);
+
+	device = g_hash_table_lookup(connection->devices,
+				     camera_source->device_id);
+
+	if (!device)
+		return;
+
+	blog(LOG_INFO, "[camera-portal] selected camera '%s'",
+	     camera_source->device_id);
+
+	camera_source->obs_pw_stream = obs_pipewire_connect_stream(
+		connection->obs_pw, camera_source->source, device->id,
+		"OBS PipeWire Camera",
+		pw_properties_new(PW_KEY_MEDIA_TYPE, "Video",
+				  PW_KEY_MEDIA_CATEGORY, "Capture",
+				  PW_KEY_MEDIA_ROLE, "Camera", NULL));
+}
+
+static bool device_selected(void *data, obs_properties_t *props,
+			    obs_property_t *property, obs_data_t *settings)
+{
+	struct camera_portal_source *camera_source = data;
+	const char *device_id;
+
+	device_id = obs_data_get_string(settings, "device_id");
+
+	if (update_device_id(camera_source, device_id))
+		stream_camera(camera_source);
+
+	return true;
+}
+
+static void populate_cameras_list(struct camera_portal_source *camera_source,
+				  obs_property_t *device_list)
+{
+	struct camera_device *device;
+	GHashTableIter iter;
+	const char *device_id;
+	bool device_found;
+
+	if (!connection)
+		return;
+
+	obs_property_list_clear(device_list);
+
+	device_found = false;
+	g_hash_table_iter_init(&iter, connection->devices);
+	while (g_hash_table_iter_next(&iter, (gpointer *)&device_id,
+				      (gpointer *)&device)) {
+		const char *device_name;
+
+		device_name = pw_properties_get(device->properties,
```

Comment:
This is a PipeWire issue, but for now, we should use more than `PW_KEY_NODE_DESCRIPTION`. This value is for my webcam "0807" for some reason. I'd suggest maybe adding the object.path Property to that string (Maybe `descr (path)`) which is sth. like `v4l2:/dev/video0`. This would also allow people to notice why a webcam used by PipeWire can't be accessed via v4l2 from another application.